### PR TITLE
Bugfix for thumbnails not rendering for s3 image URL's with spaces.

### PR DESCRIPTION
**Pull request by gregcrown**
### What does it do?
Updates file source URL's with encoded spaces.

### Why is it needed?
Media sources with spaces in URL's (example: s3) were unable to generate thumbnails.

### Related issue(s)/PR(s)
No related issues.

### CODE REVIEW

**Review by Jako — COMMENTED** [review]
urlencode or rawurlencode is not an option?

**Comment by Jako** [comment]
Will S3 media sources be deprecated in 3.x? The API has changed for a lot of new buckets and the current code does not work anymore.

**Comment by JoshuaLuckers** [comment]
> Will S3 media sources be deprecated in 3.x? The API has changed for a lot of new buckets and the current code does not work anymore.

Doesn't Flysystem cover this in 3.x?

**Comment by Jako** [comment]
Flysystem supports this, but it has to be configured.

There is an extra available for new S3 buckets.

**Comment by gregcrown** [comment]
> urlencode or rawurlencode is not an option?

I explored this option however the variables that are affected already contained full URL's. To use urlencode or rawurlencode would require breaking up the complete URL's into segments and then rebuilding the URL. Explicitly replacing the spaces removes the extra steps.

Please suggest a more elegant approach if you have one though. 👍

**Review by Jako — DISMISSED** [review]

**Comment by gregcrown** [comment]
I needed to amend my commit to include a check for URL match in:

core/model/phpthumb/phpthumb.class.php

I should have done this to begin with. This will allow filesystem images to be handled without the URL fix.

File: core/model/phpthumb/phpthumb.class.php [hunk 1]
```diff
@@ -280,6 +280,9 @@ public function purgeTempFiles() {
 	public function setSourceFilename($sourceFilename) {
 		//$this->resetObject();
 		//$this->rawImageData   = null;
+		if (preg_match('#^[a-z0-9]+://#i', $sourceFilename)) {
```
Comment (JoshuaLuckers) [inline core/model/phpthumb/phpthumb.class.php]: Please add a comment with an explanation to indicate what this does and why it's needed.

**Review by JoshuaLuckers — CHANGES_REQUESTED** [review]

File: core/model/phpthumb/phpthumb.class.php (hunk 1 above)
Comment (gregcrown) [reply core/model/phpthumb/phpthumb.class.php]: This is the same URL checking logic that exists on lines 581 and 906. Should we add comments to all of these instances as well?

Comment (JoshuaLuckers) [reply core/model/phpthumb/phpthumb.class.php]: Only here is fine with me.

Comment (Jako) [inline core/model/phpthumb/phpthumb.class.php]: That class is external code and can't be changed.

**Review by Jako — CHANGES_REQUESTED** [review]

**Comment by Ibochkarev** [comment]
@gregcrown Please continue to work on this PR. Thanks

**Comment by gregcrown** [comment]
This PR has been updated by reseting the original phpThumb and making the required changes to the MODx managed modPhpThumb class. I've also added a comment above the URL detection for clarity.

**Review by Jako — APPROVED** [review]

**Review by JoshuaLuckers — APPROVED** [review]

**Review by Mark-H — APPROVED** [review]

**Comment by Mark-H** [comment]
Hey @gregcrown, I was about to merge this but it looks like you haven't signed the MODX CLA yet. Could you please fill out the form at https://develop.modx.com/contribute/cla/ and let me know when you've done so? 

Thanks!

**Comment by gregcrown** [comment]
Accepted! Thanks for the heads up!

**Pull request merged**